src/cxx_supportlib/vendor-copy/adhoc_lve.h

```diff
@@ -0,0 +1,305 @@
+	{
+		char error_msg[8192];
+		int rc = jail_function_ptr(pw, error_msg);
+		if (rc < 0)
+			jail_err.assign(error_msg);
+		return rc;
+	}
+
+private:
+	friend class LveEnter;
+
+	/**
+	 * using boost::shared_ptr because of scoped_ptr does not allow
+	 * to specify custom destructor
+	 */
+	boost::shared_ptr<void>          liblve_handle;
+	boost::shared_ptr<struct liblve> lve_init_handle;
+
+	init_lve_function_ptr_t          init_lve_function_ptr;
+	destroy_lve_function_ptr_t       destroy_lve_function_ptr;
+	lve_enter_flags_function_ptr_t   lve_enter_flags_function_ptr;
+	lve_exit_function_ptr_t          lve_exit_function_ptr;
+	jail_function_ptr_t              jail_function_ptr;
+
+	std::string                      init_error;
+};
+
+class LveInitSignleton
+{
+public:
+	static LibLve& getInstance(std::string* outInitOneTimeError)
+	{
+		if (!instance())
+		{
+			static boost::once_flag flag = BOOST_ONCE_INIT;
+			boost::call_once(flag, initOnce, outInitOneTimeError);
+		}
+
+		return *instance();
+	}
+private:
+	typedef boost::scoped_ptr<LibLve> lvelib_scoped_ptr;
+
+	static lvelib_scoped_ptr& instance()
+	{
+		static lvelib_scoped_ptr lvelib_scoped_ptr;
+		return lvelib_scoped_ptr;
+	}
+
+	static void initOnce(std::string *err_buf)
+	{
+		instance().reset(new LibLve);
+
+		if (instance()->is_error() && err_buf)
+			*err_buf = instance()->error();
+	}
+};
+
+class LveEnter : private boost::noncopyable
+{
+public:
+	typedef void (*exit_callback_t)(bool entered, const std::string& exit_error);
+
+	LveEnter(LibLve& lve, uint32_t uid, uint32_t cfg_min_uid, exit_callback_t cb)
+		: ctx(lve)
+		, cookie(0)
+		, entered(false)
+		, exit_callback(cb)
+	{
+		enter(uid, cfg_min_uid);
+	}
+	~LveEnter()
+	{
+		exit();
+	}
+
+	LveEnter& enter(uint32_t uid, uint32_t min_uid)
+	{
+		bool is_enter_lve_allowed = (min_uid <= uid);
+
+		if (!is_enter_lve_allowed || !ctx.is_lve_ready() || entered)
+			return *this;
+
+		int err = ctx.lve_enter_flags_function_ptr(ctx.lve_init_handle.get(),
+		                                           uid,
+		                                           &cookie,
+		                                           LVE_NO_MAXENTER|LVE_SILENCE);
+		if (!err)
+			entered = true;
+		else
+			enter_exit_error << "lve_enter_flags error [" << err << "]";
+
+		return *this;
+	}
+
+	LveEnter& exit()
+	{
+		const bool memento (entered);
+
+		if (entered)
+		{
+			int err = ctx.lve_exit_function_ptr(ctx.lve_init_handle.get(), &cookie);
+			if (err)
+				enter_exit_error << "lve_exit error [" << err << "]";
+
+			entered = false;
+		}
+
+		if (exit_callback)
+			exit_callback(memento, enter_exit_error.str());
+
+		return *this;
+	}
+
+	bool is_entered() const { return entered; }
+	bool is_error() const { return !enter_exit_error.str().empty(); }
+	std::string error() const { return enter_exit_error.str(); }
+
+	LibLve& lveInstance() { return ctx; }
+	const LibLve& lveInstance() const { return ctx; }
+
+private:
+	LibLve& ctx;
```

Comment:
Please put the '&' right before the variable name, not right after the type name, per the Passenger C++ coding style:

    LibLve &ctx;